table/block_based/filter_policy.cc

```diff
@@ -203,9 +220,37 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
     return rv;
   }
 
-  // A deque avoids unnecessary copying of already-saved values
-  // and has near-minimal peak memory use.
-  std::deque<uint64_t> hash_entries_;
+  // TODO: Ideally we want to verify the hash entry
+  // as it is added to the filter and eliminate this function
+  // for speeding up and leaving fewer spaces for undetected memory/CPU
+  // corruption. For Ribbon Filter, it's bit harder.
+  // Possible solution:
+  // pass a custom iterator that tracks the xor checksum as
+  // it iterates to ResetAndFindSeedToSolve
+  Status MaybeVerifyHashEntriesChecksum() {
+    if (!detect_filter_construct_corruption_) {
+      return Status::OK();
+    }
+
+    TEST_SYNC_POINT_CALLBACK(
+        "XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum::"
+        "PreVerification",
+        &hash_entries_info_.entries);
+
+    uint64_t actual_hash_entries_xor_checksum = 0;
+    for (uint64_t h : hash_entries_info_.entries) {
+      actual_hash_entries_xor_checksum ^= h;
+    }
+
+    if (actual_hash_entries_xor_checksum == hash_entries_info_.xor_checksum) {
+      return Status::OK();
+    } else {
+      // Since these hash entries are corrupted and they will not be used
+      // anymore, we can reset them and release memory.
+      ResetEntries();
```

Comment:
I worry slightly that if someone ignores the status and proceeds with construction, we will get an "always false" (empty) filter rather than a safe "always true" filter. But I'm not sure it's worth an extra field and extra checks. Maybe just a stern warning in the comments right here. ;)

Reply:
> Maybe just a stern warning in the comments right here
Not sure I understand why we need a comment in here instead of in https://github.com/facebook/rocksdb/pull/9342/files#diff-90d330d77a09b375bb5370051e4529ee67c905d7b72b2b599de722edab9a44f7R68-R71. I can add something like "A corrupted filter should not be further used since it's unsafe". 

> But I'm not sure it's worth an extra field and extra checks.
What are those extra checks for?

Reply:
EDIT: I believe we will still get a safe "always true" filter due to  [this](https://github.com/facebook/rocksdb/pull/9342/files#diff-96639dd177bb4e9d0f353540edef15d418263a1ce94408b0b8d4cfe4711b7f1aR374). ResetEntries() is just for releasing artifacts while the returned filter should still be [AlwaysTrueFilter](https://github.com/facebook/rocksdb/pull/9342/files#diff-96639dd177bb4e9d0f353540edef15d418263a1ce94408b0b8d4cfe4711b7f1aR374). 

I don't think clearing hash entries affects the filter type - for example, we also clear hash entries before returning a normal filter when there is no corruption errr. However, I am still adding a warning to Finish API of "A corrupted filter should not be further used since it's unsafe" cuz it's good to have.